lib/plugins/lindat_auth/__init__.py

```diff
@@ -1,163 +1,242 @@
-# Copyright (c) 2016 Charles University in Prague, Faculty of Arts,
-#                    Institute of the Czech National Corpus
-# Copyright (c) 2016 Tomas Machalek <tomas.machalek @ gmail.com>
-#
-# This program is free software; you can redistribute it and/or
-# modify it under the terms of the GNU General Public License
-# as published by the Free Software Foundation; version 2
-# dated June, 1991.
-#
-# This program is distributed in the hope that it will be useful,
-# but WITHOUT ANY WARRANTY; without even the implied warranty of
-# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
-# GNU General Public License for more details.
-
-"""
-A custom authentication module for the Institute of Formal and Applied Linguistics.
-You probably want to implement an authentication solution of your own. Please refer
-to the documentation or read the dummy_auth.py module to see the required interface.
+# coding=utf-8
 """
-import crypt
-import MySQLdb
+    Authentication and authorization based on Federated login (Shibboleth) and
+    limited local user support.
+
+    This auth is not generic enough to be called ShibbolethAuth because it uses
+    specific database backed etc.
 
-from translation import ugettext as _
+"""
+import logging
+import os
 import plugins
 from plugins.abstract.auth import AbstractSemiInternalAuth
-from plugins.tree_corparch import CorptreeParser
-
+from plugins.abstract import PluginException
 
-class DbConnection(object):
+_logger = logging.getLogger(__name__)
 
-    def __init__(self, conn):
-        self.conn = conn
 
-    def get(self):
-        return self.conn
+def uni(str_str, encoding="utf-8"):
+    """ Try to get unicode without errors """
+    try:
+        if isinstance(str_str, unicode):
+            return str_str
+        elif isinstance(str_str, basestring):
+            return unicode(str_str, encoding)
+    except UnicodeError:
+        pass
+    try:
+        return unicode(str(str_str), encoding=encoding, errors='ignore')
+    except UnicodeError:
+        pass
+    return str_str.decode(encoding=encoding, errors="ignore")
 
 
-def open_db(conf):
-    conn = MySQLdb.connect(host=conf['lindat:auth_db_host'], user=conf['lindat:auth_db_username'],
-                           passwd=conf['lindat:auth_db_password'] or "", db=conf['lindat:auth_db_name'])
-    conn.set_character_set('utf8')
-    return DbConnection(conn)
-
-
-class LINDATAuth(AbstractSemiInternalAuth):
+class FederatedAuthWithFailover(AbstractSemiInternalAuth):
     """
-    A custom authentication class for the Institute of the Czech National Corpus
+        A Shibboleth authentication module with a failover
+        solution. Please note that the failover solution should be
+        used only in exceptional cases because no high security
+        measures have been applied (as opposed to federated login).
     """
+    ID_KEYS = ('HTTP_EPPN', 'HTTP_PERSISTENT_ID', 'HTTP_MAIL')
+    RESERVED_USER = '__user_count'
```

Comment:
wtf is this field used for? see other coments in the code; it's used as a list (after in keyword), it's used probably as a column name to increment db counter, it's used in string comparison

Reply:
users have to have numeric id (kontext requirement), so this is the database entry that stores a unique id